Homework DB week 3 Vladimir

intruston opened the pull request.

Breus commented on Week3/1normalization.MD:
```diff
@@ -0,0 +1,18 @@
+# 1. What columns violate 1NF?
+*food_code, food_description* contain few values;
+*dinner_date* has different date-time format;
+*member_id* is not unique, so we need to make a new primary key.
```
Correct, although `dinner_date` technically doesn't violate 1NF: it is just really ugly 😄

Breus commented on Week3/1normalization.MD:
```diff
@@ -0,0 +1,18 @@
+*member_id* is not unique, so we need to make a new primary key.
+
+# 2. What entities do you recognize that could be extracted?
+There are 4 entities: *member*, *dinner*, *venue*, *food*
+Each entities has their own ID or code and could be in independent table.
```
Correct!

Breus commented on Week3/1normalization.MD:
```diff
@@ -0,0 +1,18 @@
+
+# 3. Name all the tables and columns that would make a 3NF compliant solution.
+Let's split data on 3 tables:
+*members*  (member_id PK, member_name, member_address);
```
The thing with having dinner and venue in the same table is that it violates normalization (you repeat venue description needlessly often for multiple dinners).

There could be multiple dinners that happen at the same venue, in which case you want to refere those dinners to the same venue (row in its own table).

intruston replied:
Updated! Added table *venue* (venue_code PK, venue_description);

Breus commented on Week3/1normalization.MD:
```diff
@@ -0,0 +1,18 @@
+
+to make relations between tables we could use these one:
+*orders* (order_id PK, member_id, dinner_id);
+*goods* (id PK, dinner_id, food_code);
```
Your solution is almost correct: in the second row of the example data you see that two members (Ben and Cristina) go together to the same dinner. Therefore, you would also need to keep track in` goods`  the member who ordered that food PER dinner (so basically food PER dinner PER member). Now you just have food PER dinner in the `good` table.

Anyway, really well done! Generally this exercise is made much worse 😉

Breus commented on Week3/2transactions/transactions-create-tables.js:
```diff
@@ -0,0 +1,44 @@
+
+  const accountTable = `
+    CREATE TABLE account (
+        account_number VARCHAR(255) PRIMARY KEY,
```
Why would you have VARCHAR for a column that called `number`? Why not just INT?

Breus commented on Week3/2transactions/transactions-create-tables.js:
```diff
@@ -0,0 +1,44 @@
+
+  const accountChangesTable = `
+    CREATE TABLE account_changes (
+        change_number INT AUTO_INCREMENT PRIMARY KEY,
+        account_number VARCHAR(255) NOT NULL,
+        amount DECIMAL(10,2) NOT NULL,
+        changed_date DATETIME NOT NULL,
+        remark VARCHAR(255),
+        FOREIGN KEY (account_number) REFERENCES account (account_number)
```
Besides account_number (should be INT) it is correct!

Breus commented on Week3/2transactions/transactions-insert-values.js:
```diff
@@ -0,0 +1,52 @@
+    VALUES ?
+  `;
+
+  const insertAccountChangesData = `
+    INSERT INTO account_changes (account_number, amount, changed_date, remark)
+    VALUES ?
+  `;
+
+  connection.query(insertAccountData, [accountData], function(err, result) {
+    if (err) throw err;
+    console.log('Sample data inserted into account table');
+  });
+
+  connection.query(insertAccountChangesData, [accountChangesData], function(err, result) {
+    if (err) throw err;
+    console.log('Sample data inserted into account_changes table');
```
Nice solution and good use of prepared statements!

Breus commented on Week3/2transactions/transaction.js:
```diff
@@ -0,0 +1,76 @@
+const accountB = 'NL102';
+const changed_date = new Date().toISOString().slice(0, 19).replace('T', ' ');
+
+connection.beginTransaction((err) => {
+  if (err) throw err;
+  console.log('Transaction started');
+
+  connection.query(`SELECT balance FROM account WHERE account_number = '${accountA}'`, (err, result) => {
+
+    if (result[0].balance < transferAmount) {
+      connection.rollback(() => {
+        throw new Error("Insufficient balance");
+      });
+    }
+
+    connection.query(
+      `UPDATE account SET balance = balance - ${transferAmount} WHERE account_number = '${accountA}'`,
+      (err, result) => {
+        if (err) {
+          connection.rollback(() => {
+            throw err;
+          });
+        }
+
```
Good solution!

Breus commented on Week3/3SQLinjestion.js:
```diff
@@ -0,0 +1,47 @@
+
+      //getPopulation('country', `Netherlands' OR '1'='1`, `NLD' or '1'='1`,console.log);
+      //getPopulationNoInjection('country', `Netherlands' OR '1'='1`, `NLD' or '1'='1`,console.log);
+      getPopulationNoInjection(`country`, `Netherlands`, `NLD`,console.log);
```
Exercise 3 is correct!

Breus commented on Week3/homework/mongodb/index.js:
```diff
@@ -25,27 +47,52 @@ async function findEpisodesExercises(client) {
+  const findTitles = await client
+    .db("databaseWeek3")
+    .collection("bob_ross_episodes")
+    .find({ elements: { $in: ["CLIFF"] } })
```
You could just do `    .find({ elements: "CLIFF" })`, but what you do also works, of course

Breus commented on Week3/homework/mongodb/index.js:
```diff
@@ -78,8 +135,13 @@ async function deleteEpisodeExercise(client) {
+
   console.log(
-    `Ran a command to delete episode and it deleted ${"TODO: fill in variable here"} episodes`
+    `Ran a command to delete episode and it deleted ${deleteEP.deletedCount} episodes`
```
Exercise 4 all correct!

Breus reviewed (changes requested):
Hi @intruston, well done on your homework for week 3. I've added some remarks for exercise 1 to improve. I'm sure you will be able to figure it out!

intruston commented:
Hi @Breus . Thank you for review and sorry for a long update. I implemented changes for 1 exercise.
And added a comment about account_number VARCHAR(255) PRIMARY KEY (I did it VARCHAR because of IBAN - International Bank Account Number style) :)
